ci/ci/ci.py

```diff
@@ -121,13 +126,54 @@
 
     batch_client = request.app['batch_client']
     batches = batch_client.list_batches(
-        f'test=1 pr={pr_number}')
+        f'test=1 pr={pr.number}')
     batches = sorted([b async for b in batches], key=lambda b: b.id, reverse=True)
     page_context['history'] = [await b.status() for b in batches]
 
     return await render_template('ci', request, userdata, 'pr.html', page_context)
 
 
+async def retry_pr(wb, pr, request):
+    app = request.app
+    session = await aiohttp_session.get_session(request)
+
+    if pr.batch is None:
+        log.info('retry cannot be requested for PR #{pr.number} because it has no batch')
+        set_message(
+            session,
+            f'Retry cannot be requested for PR #{pr.number} because it has no batch.',
+            'error')
+        raise web.HTTPFound(
+            deploy_config.base_path('ci') +
+            f'/watched_branches/{wb.index}/pr/{pr.number}')
+    batch_id = pr.batch.id
+
+    dbpool = app['dbpool']
+    async with dbpool.acquire() as conn:
+        async with conn.cursor() as cursor:
+            await cursor.execute('INSERT INTO invalidated_batches (batch_id) VALUES (%s);', batch_id)
+
+    pr.batch = None
+    pr.build_state = None
+    await wb.notify_batch_changed(app)
+
+    log.info(f'retry requested for PR: {pr.number}')
+    set_message(session, f'Retry requested for PR #{pr.number}.', 'info')
+
+
+@routes.post('/watched_branches/{watched_branch_index}/pr/{pr_number}/retry')
+@check_csrf_token
+@web_authenticated_developers_only(redirect=False)
+async def post_retry_pr(request, userdata):  # pylint: disable=unused-argument
+    wb, pr = wb_and_pr_from_request(request)
+
+    await asyncio.shield(retry_pr(wb, pr, request))
+
+    raise web.HTTPFound(
+        deploy_config.base_path('ci') +
```

Comment:
I think I had intended `url` for this instead base_path + of explicit string concatenation, although that will get you an absolute URL (which I prefer?)

Reply:
My gut says absolute links are bad practice but I have no reasoning. I'll switch to url.